File: sdk/java/picodroid/util/Log.java
// SPDX-License-Identifier: GPL-3.0-only
package picodroid.util;

/**
 * API for sending log output, mirroring {@code android.util.Log}: the {@link #v}/{@link #d}/{@link
 * #i}/{@link #w}/{@link #e} severity ladder plus {@link #wtf}, each with a {@code Throwable}
 * overload that appends the throwable's message.
 *
 * <p>On hardware the level maps to the matching defmt severity (RTT viewers can filter on it); the
 * simulator prints every level in the same {@code [Tag] message} line format.
 */
public class Log {
  public static native void v(String tag, String msg);

  public static native void d(String tag, String msg);

  public static native void i(String tag, String msg);

  public static native void w(String tag, String msg);

  public static native void e(String tag, String msg);

  /**
   * What a Terrible Failure: a condition that should never happen. Logs at error severity —
   * picodroid has no separate ASSERT channel.
   */
  public static void wtf(String tag, String msg) {
    e(tag, msg);
  }

  public static void v(String tag, String msg, Throwable tr) {
    v(tag, append(msg, tr));
  }

  public static void d(String tag, String msg, Throwable tr) {
    d(tag, append(msg, tr));
  }

  public static void i(String tag, String msg, Throwable tr) {
    i(tag, append(msg, tr));
  }

  public static void w(String tag, String msg, Throwable tr) {
    w(tag, append(msg, tr));
  }

  public static void e(String tag, String msg, Throwable tr) {
    e(tag, append(msg, tr));
  }

  public static void wtf(String tag, String msg, Throwable tr) {
    e(tag, append(msg, tr));
  }

  private static String append(String msg, Throwable tr) {
    if (tr == null) {
      return msg;
    }
    String m = tr.getMessage();
    return m == null ? msg : msg + ": " + m;
  }
}

File: picodroid-core/src/util/log.rs
// SPDX-License-Identifier: GPL-3.0-only
use pico_jvm::{
    heap::StringTable,
    types::{JvmError, Value},
};

/// Severity of a `picodroid.util.Log` call. Maps onto defmt's level ladder on
/// hardware; the simulator prints every level in the same `[Tag] message`
/// stdout format so sim-test log-token greps stay level-agnostic.
#[derive(Copy, Clone)]
pub enum LogLevel {
    Verbose,
    Debug,
    Info,
    Warn,
    Error,
}

/// Native implementation of `picodroid.util.Log.{v,d,i,w,e}(String tag, String msg)`.
/// Resolves both string references and emits a log line at `level`.
pub fn log(level: LogLevel, args: &[Value], strings: &StringTable) -> Result<(), JvmError> {
    let tag = resolve(args.first().copied().unwrap_or(Value::Null), strings)?;
    let msg = resolve(args.get(1).copied().unwrap_or(Value::Null), strings)?;
    #[cfg(not(feature = "sim"))]
    match level {
        LogLevel::Verbose => defmt::trace!("{=str}: {=str}", tag, msg),
        LogLevel::Debug => defmt::debug!("{=str}: {=str}", tag, msg),
        LogLevel::Info => defmt::info!("{=str}: {=str}", tag, msg),
        LogLevel::Warn => defmt::warn!("{=str}: {=str}", tag, msg),
        LogLevel::Error => defmt::error!("{=str}: {=str}", tag, msg),
    }
    #[cfg(feature = "sim")]
    {
        let _ = level;
        println!("[{}] {}", tag, msg);
    }
    Ok(())
}

fn resolve(v: Value, strings: &StringTable) -> Result<&str, JvmError> {
    match v {
        Value::Reference(idx) => strings.resolve(idx).ok_or(JvmError::InvalidReference),
        _ => Err(JvmError::InvalidReference),
    }
}
